lightning/src/routing/network_graph.rs

```diff
@@ -0,0 +1,1640 @@
+}
+
+impl NetworkGraph {
+	/// Returns a list of known valid channels
+	pub fn get_channels<'a>(&'a self) -> &'a BTreeMap<u64, ChannelInfo> { &self.channels }
+	/// Returns a mutable reference to an existing channel
+	pub fn get_channel_mut<'a>(&'a mut self, chan_id: &u64) -> Option<&mut ChannelInfo> { self.channels.get_mut(chan_id) }
+	/// Store a valid channel
+	pub fn add_channel(&mut self, chan_id: u64, channel_info: ChannelInfo) {
+		self.channels.insert(chan_id, channel_info);
+	}
+	/// Returns a list of known nodes
+	pub fn get_nodes<'a>(&'a self) -> &'a BTreeMap<PublicKey, NodeInfo> { &self.nodes }
+	/// Returns a mutable reference to an existing node
+	pub fn get_node_mut<'a>(&'a mut self, node_id: &PublicKey) -> Option<&mut NodeInfo> { self.nodes.get_mut(node_id) }
+	/// Store a valid node
+	pub fn add_node(&mut self, node_id: PublicKey, node_info: NodeInfo) {
+		self.nodes.insert(node_id, node_info);
+	}
+}
+
+impl Writeable for NetworkGraph {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		(self.channels.len() as u64).write(writer)?;
+		for (ref chan_id, ref chan_info) in self.channels.iter() {
+			(*chan_id).write(writer)?;
+			chan_info.write(writer)?;
+		}
+		(self.nodes.len() as u64).write(writer)?;
+		for (ref node_id, ref node_info) in self.nodes.iter() {
+			node_id.write(writer)?;
+			node_info.write(writer)?;
+		}
+		Ok(())
+	}
+}
+
+impl Readable for NetworkGraph {
+	fn read<R: ::std::io::Read>(reader: &mut R) -> Result<NetworkGraph, DecodeError> {
+		let channels_count: u64 = Readable::read(reader)?;
+		let mut channels = BTreeMap::new();
+		for _ in 0..channels_count {
+			let chan_id: u64 = Readable::read(reader)?;
+			let chan_info = Readable::read(reader)?;
+			channels.insert(chan_id, chan_info);
+		}
+		let nodes_count: u64 = Readable::read(reader)?;
+		let mut nodes = BTreeMap::new();
+		for _ in 0..nodes_count {
+			let node_id = Readable::read(reader)?;
+			let node_info = Readable::read(reader)?;
+			nodes.insert(node_id, node_info);
+		}
+		Ok(NetworkGraph {
+			channels,
+			nodes,
+		})
+	}
+}
+
+impl std::fmt::Display for NetworkGraph {
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+		write!(f, "Network map\n[Channels]\n")?;
+		for (key, val) in self.channels.iter() {
+			write!(f, " {}: {}\n", key, val)?;
+		}
+		write!(f, "[Nodes]\n")?;
+		for (key, val) in self.nodes.iter() {
+			write!(f, " {}: {}\n", log_pubkey!(key), val)?;
+		}
+		Ok(())
+	}
+}
+
+impl NetworkGraph {
+	#[inline]
+	/// Key used to link channel on the Bitcoin chain to other entities
+	pub fn get_key(short_channel_id: u64, _: Sha256dHash) -> u64 {
+		short_channel_id
+	}
+
+	#[inline]
+	/// Key used to link channel on the Bitcoin chain to other entities
+	pub fn get_short_id(id: &u64) -> &u64 {
+		id
+	}
+}
+
+impl RoutingMessageHandler for NetGraphMsgHandler {
+
+	fn handle_node_announcement(&self, msg: &msgs::NodeAnnouncement) -> Result<bool, LightningError> {
```

Comment:
I feel like if we probably want these to be in NetworkGraph, not NetGraphMsgHandler and just have thing wrappers across NetGraphMsgHandler. At a high level, the goal of encapsulating NetworkGraph and not letting users simply have direct write access to all the internal fields is that they have to stay consistent - by allowing someone to, eg, take a NodeAnnouncement, pass it into a NetworkGraph (with &mut self), we can easily provide the ability to update NetworkGraph without leaving us in an inconsistent state.